Doc/glossary.rst

```diff
@@ -363,6 +400,24 @@ Glossary
       the :term:`cyclic garbage collector <garbage collection>` is to identify these groups and break the reference
       cycles so that the memory can be reclaimed.
 
+   data race
+      A situation where multiple threads access the same memory location
+      concurrently, at least one of the accesses is a write, and the threads
+      do not use any synchronization to control their access.  Data races
+      lead to :term:`non-deterministic` behavior and can cause data corruption.
+      Proper use of :term:`locks <lock>` and other :term:`synchronization primitives
+      <synchronization primitive>` prevents data races.  See also
+      :term:`race condition` and :term:`thread-safe`.
+
+   deadlock
+      A situation where two or more threads are unable to proceed because
+      each is waiting for the other to release a resource.  For example,
+      if thread A holds lock 1 and waits for lock 2, while thread B holds
+      lock 2 and waits for lock 1, both threads will wait indefinitely.
+      Deadlocks can be avoided by always acquiring multiple :term:`locks <lock>`
+      in a consistent order or by using timeout-based locking.  See also
+      :term:`lock` and :term:`reentrant`.
```

Comment:
Maybe add "Any program that makes blocking calls using more than one lock is possibly susceptible to deadlocks."